File: reverse-integer.cpp
/* https://leetcode.com/problems/reverse-integer
 * Solution: take care of edge cases where it overflows for positive and negative integers
*/

class Solution {
public:
    int reverse(int x) {
        int res = 0;
        while(x!=0){
            if((long long)res*10>INT_MAX || (long long)res*10<INT_MIN) return 0;
            else res=res*10 + x%10;
            x/=10;
        }
        return res;
    }
};

File: 7/reverse-integer.go
func reverse(x int) int {
	rev, md := 0, 0
	int_max, int_min := 1<<31-1, -1<<31

	for x != 0 {
		md = x % 10
		if rev > int_max/10 || (rev == int_max/10 && md > 7) {
			return 0
		}
		if rev < int_min/10 || (rev == int_min/10 && md < -8) {
			return 0
		}

		rev = rev*10 + md
		x /= 10
	}

	return rev
}
